src/state/slices/portfolioSlice/portfolioSlice.ts

```diff
@@ -22,15 +22,29 @@ export const portfolio = createSlice({
       moduleLogger.info('clearing portfolio')
       return initialState
     },
+    setWalletId: (state, { payload }: { payload: WalletId | undefined }) => {
+      // note this function can unset the walletId to undefined
+      moduleLogger.info(payload, 'setting wallet id')
+      state.walletId = payload
+      if (!payload) return
+      state.wallet.ids = Array.from(new Set([...state.wallet.ids, payload]))
+    },
     upsertAccountMetadata: (state, { payload }: { payload: AccountMetadataById }) => {
       moduleLogger.debug('upserting account metadata')
       state.accountMetadata.byId = {
         ...state.accountMetadata.byId,
         ...payload,
       }
-      state.accountMetadata.ids = Array.from(
-        new Set([...state.accountMetadata.ids, ...Object.keys(payload)]),
-      )
+      const accountIds = Object.keys(payload)
+      state.accountMetadata.ids = Array.from(new Set([...state.accountMetadata.ids, ...accountIds]))
+
+      if (!state.walletId) return
+      const existingWalletAccountIds = state.wallet.byId[state.walletId] ?? []
```

Comment:
when upserting an account, index it against the wallet, so when the wallet returns, we know all the account ids without reaching out over the network